### PR TITLE
windows: move ffmpeg bins to submodule

### PR DESCRIPTION
**Pull request by shuffle2**
windows: move ffmpeg bins to submodule
udpate ffmpeg to 28d011516b72a2401b2289a4be34c6e8ab611a0b
enable ffmpeg on arm64
plumb ffmpeg logging to dolphin logging

contains the change from #10249 to enable utvideo, since that was the motivation to fixup ffmpeg stuff.

I also experimented with adding h264/hevc encoders (the Media Foundation backends are nice), but realized users that want that sort of functionality will probably just use xbox game bar recording or something.

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Yeah, I agree this is the better way to do this.

**Comment by shuffle2** [comment]
it seems to make git mad when a directory is removed and a submodule is created in the same location in the same commit.
i'm going to try breaking it into two commits. If that doesn't work, might have to name the submodule something else

**Comment by JosJuice** [comment]
Are you sure the problem isn't that you didn't append `.git` to the URL?

**Comment by shuffle2** [comment]
> Are you sure the problem isn't that you didn't append `.git` to the URL?

I don't think it's a problem? The submodule works in a local clone of the shuffle2/dolphin repo. But there is a problem like:
```
>git checkout win-ffmpeg
Switched to a new branch 'win-ffmpeg'
Branch 'win-ffmpeg' set up to track remote branch 'win-ffmpeg' from 'origin'.
>git submodule update
Cloning into 'C:/src/dolphintest/Externals/FFmpeg'...
starting fsmonitor-daemon in 'C:/src/dolphintest/Externals/FFmpeg'
Submodule path 'Externals/FFmpeg': checked out '8cc3c8847438521c89e6e84b349009978e58f8b5'
>git checkout master
warning: unable to rmdir 'Externals/FFmpeg': Directory not empty
```
...git leaves Externals/FFmpeg with submodule contents instead of deleting it. I wonder if that is the problem?

**Comment by shuffle2** [comment]
hrm, after looking at the FS on the windows buildbot, it seems .git/modules/Externals/ffmpeg is the content of the submodule introduced by #10249 (case insensitive FS, remember...). Maybe that is the problem.

File: Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj [hunk 1]
```diff
@@ -18,8 +18,9 @@
   <ItemDefinitionGroup>
     <Link>
       <AdditionalDependencies>avrt.lib;iphlpapi.lib;winmm.lib;setupapi.lib;rpcrt4.lib;comctl32.lib;Shlwapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
-      <AdditionalDependencies Condition="'$(Platform)'=='x64'">opengl32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;%(AdditionalDependencies)</AdditionalDependencies>
-      <AdditionalLibraryDirectories Condition="'$(Platform)'=='x64'">$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalDependencies Condition="'$(Platform)'=='x64'">opengl32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>libavcodec.a;libavformat.a;libavutil.a;libswresample.a;libswscale.a;Bcrypt.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (AdmiralCurtiss) [inline Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj:22]: `.a`? Seems like that's correct according to the submodule, but that's rather unusual for Windows.

File: Source/Core/VideoCommon/FrameDump.cpp [hunk 2]
```diff
@@ -69,9 +70,6 @@ void InitAVCodec()
   static bool first_run = true;
   if (first_run)
   {
-#if LIBAVCODEC_VERSION_MICRO >= 100 && LIBAVCODEC_VERSION_INT < AV_VERSION_INT(58, 9, 100)
-    av_register_all();
-#endif
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/FrameDump.cpp:74]: Why did you remove this? Do versions that are old enough to require this not compile anyway, making this check pointless?

File: Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj (hunk 1 above)
Comment (shuffle2) [reply Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj:22]: the `configure` script of ffmpeg chooses this filename format when built under wsl. The file content is the same. We could rename it but I'm not aware it's a problem.

File: Source/Core/VideoCommon/FrameDump.cpp (hunk 2 above)
Comment (shuffle2) [reply Source/Core/VideoCommon/FrameDump.cpp:74]: This `#if` was added to maintain compat with the version of ffmpeg in Externals before. It shouldn't be required anymore.

File: Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj (hunk 1 above)
Comment (AdmiralCurtiss) [reply Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj:22]: Fair enough.

File: Source/Core/VideoCommon/FrameDump.cpp (hunk 2 above)
Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/FrameDump.cpp:74]: Ah, okay. Probably fine to remove it then.

File: Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj (hunk 1 above)
Comment (shuffle2) [reply Source/Core/DolphinNoGUI/DolphinNoGUI.vcxproj:22]: fwiw, it turns out to be a problem with dolphin's existing `CMake/FindFFmpeg.cmake`, and i'd rather just rename the files to be more windows-y than fiddle with the cmake

File: Source/Core/VideoCommon/FrameDump.cpp [hunk 3]
```diff
@@ -69,11 +73,33 @@ void InitAVCodec()
...
+    // av_log_set_level(AV_LOG_DEBUG);
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/FrameDump.cpp:76]: Did you intentionally leave this commented out?

File: Source/Core/VideoCommon/FrameDump.cpp [hunk 4]
```diff
@@ -215,7 +248,16 @@ bool FrameDump::CreateVideoFile()
...
+
+  if (m_context->codec->codec_id == AV_CODEC_ID_FFV1)
+    m_context->codec->pix_fmt = AV_PIX_FMT_BGR0;
+  else if (m_context->codec->codec_id == AV_CODEC_ID_UTVIDEO)
+  {
+    m_context->codec->pix_fmt = AV_PIX_FMT_GBRP;
+    av_opt_set_int(m_context->codec->priv_data, "pred", 3, 0);  // median
+  }
+  else
+    m_context->codec->pix_fmt = AV_PIX_FMT_YUV420P;
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/FrameDump.cpp:260]: If one branch uses braces, all of them should.

File: Source/Core/VideoCommon/FrameDump.cpp (hunk 3 above)
Comment (shuffle2) [reply Source/Core/VideoCommon/FrameDump.cpp:76]: Yes, it defaults to some level not as verbose as debug ( although in my short testing even debug doesn’t really spam a lot). I figured I’d leave the comment for people debugging in the future but it’s probably fine to commit uncommented- I don’t have strong opinion either way

Comment (shuffle2) [reply Source/Core/VideoCommon/FrameDump.cpp:76]: i'll just uncomment it 🤷‍♂️

**Comment by AdmiralCurtiss** [comment]
So is this good to go? Looks fine to me.

**Comment by shuffle2** [comment]
from my side it is

**Comment by JosJuice** [comment]
@shuffle2 I would like to copy the contents of https://github.com/shuffle2/ext-win-ffmpeg to https://github.com/dolphin-emu/ext-win-ffmpeg before merging this, so that building Dolphin doesn't rely on the GitHub accounts of individuals still existing. Does this sound fine to you? I would create the repo and you would have to change the `.gitmodules` file in this PR.

**Comment by shuffle2** [comment]
> @shuffle2 I would like to copy the contents of https://github.com/shuffle2/ext-win-ffmpeg to https://github.com/dolphin-emu/ext-win-ffmpeg before merging this, so that building Dolphin doesn't rely on the GitHub accounts of individuals still existing. Does this sound fine to you? I would create the repo and you would have to change the `.gitmodules` file in this PR.

yes that's fine

**Comment by JosJuice** [comment]
It has now been created at https://github.com/dolphin-emu/ext-win-ffmpeg.

I'm not sure exactly how the permissions are set up. I'm assuming there isn't more you need to do inside that repository for now, but if you want to do something with the repository in the future and the permissions seem to restrictive, just let me know.

**Comment by shuffle2** [comment]
it's switched over, and i checked that windows buildbot switched to the correct remote for the submodule as well (it did; didn't need any intervention)

**Comment by AdmiralCurtiss** [comment]
Tested, works. Let's merge this!
